File: apps/ios/Sources/Models/TurnActivityModel.swift
import Foundation

/// Static + dynamic descriptors for the Turn Live Activity.
///
/// `TurnActivityAttributes` is split into two pieces by the same pattern
/// ActivityKit's `ActivityAttributes` requires:
///   - the "attributes" themselves (`agentName`, `sessionID`) are
///     immutable for the lifetime of the activity,
///   - `ContentState` carries everything that mutates during the turn
///     (current tool, elapsed time, status).
///
/// The pure-data shape lives here, separate from the ActivityKit-shaped
/// types in `TurnLiveActivityController.swift`, so the state machine that
/// decides when to start/update/end an activity can be exercised in unit
/// tests without paying the cost of importing ActivityKit (which doesn't
/// behave well under XCTest on the simulator without a registered widget
/// extension).
public struct TurnActivityAttributesData: Equatable, Hashable, Codable, Sendable {
    public var agentName: String
    public var sessionID: String

    public init(agentName: String, sessionID: String) {
        self.agentName = agentName
        self.sessionID = sessionID
    }
}

/// The mutable per-turn state mirrored into the lock-screen card.
public struct TurnActivityContentState: Equatable, Hashable, Codable, Sendable {
    public var currentTool: String?
    public var currentCommand: String?
    public var startedAt: Date
    public var tokensIn: Int
    public var tokensOut: Int
    /// "running", "pending", or "exited" — matches Litter's vocabulary
    /// so the widget renderer can switch on a known string set.
    public var status: String

    public init(
        currentTool: String? = nil,
        currentCommand: String? = nil,
        startedAt: Date,
        tokensIn: Int = 0,
        tokensOut: Int = 0,
        status: String = "running"
    ) {
        self.currentTool = currentTool
        self.currentCommand = currentCommand
        self.startedAt = startedAt
        self.tokensIn = tokensIn
        self.tokensOut = tokensOut
        self.status = status
    }
}

/// Lightweight projection of `ConversationItem` carrying just the fields
/// the state machine needs. Keeps the model free of any Rust-core types
/// so the tests don't have to depend on the generated UniFFI module.
public struct TurnActivityItem: Equatable, Hashable, Sendable {
    public enum Kind: String, Sendable {
        case tool
        case command
        case message
        case exit
        case other
    }

    public var id: String
    public var kind: Kind
    public var toolName: String?
    public var command: String?
    public var status: String
    public var exitCode: Int32?
    public var timestamp: Date

    public init(
        id: String,
        kind: Kind,
        toolName: String? = nil,
        command: String? = nil,
        status: String = "running",
        exitCode: Int32? = nil,
        timestamp: Date
    ) {
        self.id = id
        self.kind = kind
        self.toolName = toolName
        self.command = command
        self.status = status
        self.exitCode = exitCode
        self.timestamp = timestamp
    }
}

/// Side effect a `TurnActivityModel` step wants the controller to apply.
///
/// `start` / `update` / `end` map 1:1 to the three ActivityKit verbs.
/// The controller is responsible for actually calling
/// `Activity.request` / `activity.update` / `activity.end`; the model is
/// pure data so the state machine is unit-testable.
public enum TurnActivityEffect: Equatable, Sendable {
    case noop
    case start(attributes: TurnActivityAttributesData, state: TurnActivityContentState)
    case update(state: TurnActivityContentState)
    case end(state: TurnActivityContentState)
}

/// Pure state machine that decides whether to start / update / end the
/// Live Activity for a single session.
///
/// **Transitions** (mirrors Litter's `TurnLiveActivityController`):
///   - first `.tool` or `.command` item arrives → emit `.start`
///   - subsequent `.tool` / `.command` items for the active turn → `.update`
///   - `.exit` item or `status == "exited"` on the active item → `.end`
///   - tick-driven `.end` after `idleTimeout` of no tool/command activity
///   - session-exit signal from outside → `.end`
///
/// The model owns the *current* `agentName` + `sessionID` once activity
/// has started; the controller resets the model when a different session
/// becomes active.
public struct TurnActivityModel: Equatable, Sendable {
    /// Idle window after the last tool/command before the activity is ended.
    /// 5 s mirrors the spec — tuned so a chain of fast tool calls keeps
    /// the activity alive without flashing on/off between back-to-back tools.
    public static let defaultIdleTimeout: TimeInterval = 5

    public private(set) var attributes: TurnActivityAttributesData?
    public private(set) var contentState: TurnActivityContentState?
    public private(set) var lastActivityAt: Date?
    public private(set) var idleTimeout: TimeInterval

    public var isActive: Bool { attributes != nil && contentState != nil }

    public init(idleTimeout: TimeInterval = TurnActivityModel.defaultIdleTimeout) {
        self.idleTimeout = idleTimeout
    }

    /// Apply a new conversation item to the state machine and return the
    /// effect the controller should perform. `agentName` is captured at
    /// `start` time and not re-read after — the activity carries the
    /// agent that owned the turn even if the session switches.
    public mutating func apply(
        item: TurnActivityItem,
        sessionID: String,
        agentName: String
    ) -> TurnActivityEffect {
        // Terminal kinds end the activity unconditionally.
        if item.kind == .exit {
            return endActivity(at: item.timestamp, status: "exited")
        }
        if item.status == "exited", contentState != nil {
            return endActivity(at: item.timestamp, status: "exited")
        }

        // Only tool/command items drive start/update — chat messages alone
        // don't justify a lock-screen card.
        guard item.kind == .tool || item.kind == .command else {
            return .noop
        }

        lastActivityAt = item.timestamp

        if !isActive {
            let attrs = TurnActivityAttributesData(agentName: agentName, sessionID: sessionID)
            let state = TurnActivityContentState(
                currentTool: item.toolName,
                currentCommand: item.command,
                startedAt: item.timestamp,
                tokensIn: 0,
                tokensOut: 0,
                status: "running"
            )
            attributes = attrs
            contentState = state
            return .start(attributes: attrs, state: state)
        }

        // Same session: produce an update with the new tool/command.
        var next = contentState ?? TurnActivityContentState(startedAt: item.timestamp)
        next.currentTool = item.toolName ?? next.currentTool
        next.currentCommand = item.command ?? next.currentCommand
        next.status = "running"
        contentState = next
        return .update(state: next)
    }

    /// Externally-signalled end (session lifecycle exit, controller reset,
    /// app teardown). Idempotent — calling on an inactive model is a noop.
    public mutating func sessionExited(at when: Date = Date(), status: String = "exited") -> TurnActivityEffect {
        return endActivity(at: when, status: status)
    }

    /// Time-driven step. Called periodically (or on any external nudge) so
    /// the model can end the activity after `idleTimeout` without a new tool.
    /// Returns `.end` exactly once per active turn.
    public mutating func tick(now: Date) -> TurnActivityEffect {
        guard let last = lastActivityAt, isActive else { return .noop }
        if now.timeIntervalSince(last) >= idleTimeout {
            return endActivity(at: now, status: "exited")
        }
        return .noop
    }

    /// Apply a fresh token-count update without changing the active tool.
    /// Token counts arrive on a different channel from tool calls, so
    /// this is a separate input.
    public mutating func updateTokens(tokensIn: Int, tokensOut: Int) -> TurnActivityEffect {
        guard var next = contentState else { return .noop }
        next.tokensIn = tokensIn
        next.tokensOut = tokensOut
        contentState = next
        return .update(state: next)
    }

    private mutating func endActivity(at when: Date, status: String) -> TurnActivityEffect {
        guard var final = contentState else {
            attributes = nil
            return .noop
        }
        final.status = status
        let effect = TurnActivityEffect.end(state: final)
        attributes = nil
        contentState = nil
        lastActivityAt = nil
        return effect
    }
}

File: apps/android/app/src/main/kotlin/sh/nikhil/swekitty/widget/TurnActivityModel.kt
package sh.nikhil.swekitty.widget

/**
 * Android mirror of iOS `apps/ios/Sources/Models/TurnActivityModel.swift`
 * (PLAN-2026-05-19 Package 7).
 *
 * Pure state machine that decides whether to **start** / **update** /
 * **end** an "active turn" surface for a single session. iOS feeds the
 * resulting effects into ActivityKit (`Activity.request` /
 * `activity.update` / `activity.end`); Android wires them into a
 * Glance widget update in a follow-up PR. This file ships the pure
 * shape + tests so the two platforms can't drift on the transition
 * rules.
 *
 * The data shapes (`TurnActivityAttributesData`, `ContentState`,
 * `TurnActivityItem`) match the Swift types verbatim — same field
 * names, same value semantics — so a future shared-`core` move can
 * lift either side into the Rust bridge without churning the
 * platform call sites.
 */

/**
 * Static descriptors for the activity. Pinned at start time and
 * unchanged for the lifetime of the active turn. Mirrors iOS
 * `TurnActivityAttributesData`.
 */
data class TurnActivityAttributesData(
    val agentName: String,
    val sessionID: String,
)

/**
 * Mutable per-turn state. Mirrors iOS `TurnActivityContentState`.
 *
 * `startedAt` is epoch millis (Swift uses `Date`; we use `Long` so
 * the model stays serializable to / from a plain WorkData / Glance
 * preferences). `status` matches Litter's vocabulary — `"running"`,
 * `"pending"`, `"exited"`.
 */
data class TurnActivityContentState(
    val currentTool: String? = null,
    val currentCommand: String? = null,
    val startedAtMillis: Long,
    val tokensIn: Int = 0,
    val tokensOut: Int = 0,
    val status: String = "running",
)

/**
 * Lightweight projection of `ConversationItem`. Carries just the
 * fields the state machine cares about so tests don't have to depend
 * on the generated UniFFI module. Mirrors iOS `TurnActivityItem`.
 */
data class TurnActivityItem(
    val id: String,
    val kind: Kind,
    val toolName: String? = null,
    val command: String? = null,
    val status: String = "running",
    val exitCode: Int? = null,
    val timestampMillis: Long,
) {
    enum class Kind { TOOL, COMMAND, MESSAGE, EXIT, OTHER }
}

/**
 * Side effect a `TurnActivityModel` step wants the host to apply.
 * `START` / `UPDATE` / `END` map 1:1 to the three iOS ActivityKit
 * verbs. Android's host (in a follow-up PR) maps them to:
 *   - `START` → `GlanceAppWidgetManager.update(...)` first emit
 *   - `UPDATE` → subsequent `update(...)` calls
 *   - `END` → `update(...)` with empty / "Idle" state
 */
sealed class TurnActivityEffect {
    data object Noop : TurnActivityEffect()
    data class Start(
        val attributes: TurnActivityAttributesData,
        val state: TurnActivityContentState,
    ) : TurnActivityEffect()
    data class Update(val state: TurnActivityContentState) : TurnActivityEffect()
    data class End(val state: TurnActivityContentState) : TurnActivityEffect()
}

/**
 * Pure state machine for a single session. Mutates internal state and
 * returns one [TurnActivityEffect] per `apply` / `tick` call.
 *
 * Transition rules (verbatim port of iOS):
 *  - First `TOOL` or `COMMAND` item arrives → `Start`
 *  - Subsequent `TOOL` / `COMMAND` items → `Update`
 *  - `EXIT` item OR an item with `status == "exited"` → `End`
 *  - Tick after [idleTimeoutMillis] with no fresh tool item → `End`
 *  - Explicit [sessionExited] call → `End`
 */
class TurnActivityModel(
    val idleTimeoutMillis: Long = DEFAULT_IDLE_TIMEOUT_MILLIS,
) {
    var attributes: TurnActivityAttributesData? = null
        private set
    var contentState: TurnActivityContentState? = null
        private set
    var lastActivityAtMillis: Long? = null
        private set

    val isActive: Boolean
        get() = attributes != null && contentState != null

    fun apply(
        item: TurnActivityItem,
        sessionID: String,
        agentName: String,
    ): TurnActivityEffect {
        // Terminal kinds end the activity unconditionally.
        if (item.kind == TurnActivityItem.Kind.EXIT) {
            return endActivity(item.timestampMillis, status = "exited")
        }
        if (item.status == "exited" && contentState != null) {
            return endActivity(item.timestampMillis, status = "exited")
        }

        // Only TOOL / COMMAND drive start/update — bare MESSAGE rows
        // don't justify a widget update.
        if (item.kind != TurnActivityItem.Kind.TOOL && item.kind != TurnActivityItem.Kind.COMMAND) {
            return TurnActivityEffect.Noop
        }

        lastActivityAtMillis = item.timestampMillis

        if (!isActive) {
            val attrs = TurnActivityAttributesData(agentName = agentName, sessionID = sessionID)
            val state = TurnActivityContentState(
                currentTool = item.toolName,
                currentCommand = item.command,
                startedAtMillis = item.timestampMillis,
                tokensIn = 0,
                tokensOut = 0,
                status = "running",
            )
            attributes = attrs
            contentState = state
            return TurnActivityEffect.Start(attrs, state)
        }

        val next = contentState!!.copy(
            currentTool = item.toolName ?: contentState!!.currentTool,
            currentCommand = item.command ?: contentState!!.currentCommand,
            status = "running",
        )
        contentState = next
        return TurnActivityEffect.Update(next)
    }

    /**
     * Externally-signalled end (session lifecycle exit, host reset,
     * app teardown). Idempotent: calling on an inactive model is a
     * Noop.
     */
    fun sessionExited(
        nowMillis: Long = System.currentTimeMillis(),
        status: String = "exited",
    ): TurnActivityEffect = endActivity(nowMillis, status)

    /**
     * Time-driven step. Called periodically (or on any external
     * nudge) so the model can end the activity after [idleTimeoutMillis]
     * without a new tool. Returns `End` exactly once per active turn.
     */
    fun tick(nowMillis: Long): TurnActivityEffect {
        val last = lastActivityAtMillis ?: return TurnActivityEffect.Noop
        if (!isActive) return TurnActivityEffect.Noop
        return if (nowMillis - last >= idleTimeoutMillis) {
            endActivity(nowMillis, status = "exited")
        } else {
            TurnActivityEffect.Noop
        }
    }

    /**
     * Apply a fresh token-count update without changing the active
     * tool. Token counts arrive on a different channel from tool
     * calls, so this is a separate input.
     */
    fun updateTokens(tokensIn: Int, tokensOut: Int): TurnActivityEffect {
        val cur = contentState ?: return TurnActivityEffect.Noop
        val next = cur.copy(tokensIn = tokensIn, tokensOut = tokensOut)
        contentState = next
        return TurnActivityEffect.Update(next)
    }

    private fun endActivity(whenMillis: Long, status: String): TurnActivityEffect {
        val cur = contentState
        if (cur == null) {
            attributes = null
            return TurnActivityEffect.Noop
        }
        val final = cur.copy(status = status)
        val effect = TurnActivityEffect.End(final)
        attributes = null
        contentState = null
        lastActivityAtMillis = null
        return effect
    }

    companion object {
        /**
         * Idle window after the last tool/command before the activity
         * is ended. 5 seconds mirrors iOS — tuned so a chain of fast
         * tool calls keeps the activity alive without flashing on/off.
         */
        const val DEFAULT_IDLE_TIMEOUT_MILLIS: Long = 5_000L
    }
}
